lib/1.4/utility.dml

```diff
@@ -192,15 +192,105 @@ template _init_val_power_on_reset is (init_val, power_on_reset) {
     }
 }
 
+/**
+### signal_port
+
+<section>
+
+#### Description
+
+Implements a signal interface with saved state. The current state of the signal
+is stored in the saved boolean `high` and an error is logged if the signal is
+raised or lowered when already high or low. The methods `signal_raise` and
+`signal_lower` can be overridden to add additional side effects.
+
+#### Related Templates
+
+poreset, hreset, sreset
+
+</section>
+*/
+template signal_port {
+    implement signal {
+        saved bool high;
+        method signal_raise() default {
+            if (high)
+                log error: "%s raised when already high", qname;
+            else
+                high = true;
+        }
+        method signal_lower() default {
+            if (!high)
+                log error: "%s lowered when already low", qname;
+            else
+                high = false;
+        }
+    }
```

Comment:
As seen in your original connect code below, also experienced modelers can construct a signal connect that calls raise or lower early during checkpoint restore. If this happens before the `high` attribute is restored, then the tattle logic will be broken.

I can't immediately see any realistic cases where this would lead to false errors being reported; however, if a connect (incorrectly) raises the signal upon checkpoint restore, then the (righteous) error message will happen or not depending on which object gets it attributes restored first, and nondeterministic errors are unfriendly.

Reply:
Indeed, if hot-plugging must be supported it becomes quite tricky.
Considering that hot-plugging is quite rare, especially for `signal` we could make the template not support it.
On the other hand, since the template becomes a bit pointless without it, since the basic use case without hot plugging is almost trivial.